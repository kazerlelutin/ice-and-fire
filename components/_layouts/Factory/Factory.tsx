/* eslint-disable react/display-name */
export default function Factory<T>(hook: Function, Component: any):Function {
  return (props: T): JSX.Element => <Component {...props} {...hook(props)} />
}