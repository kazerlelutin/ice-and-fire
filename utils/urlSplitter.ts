//for extract id 
export default function urlSplitter(url:string):string {
  const splitUrl = url.split('/')
  return splitUrl.at(-1)
}