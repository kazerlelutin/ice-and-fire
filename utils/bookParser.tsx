import { IapiBook } from "../types/apiBook.interface";
import { IBook } from "../types/book.interface";
import urlSplitter from "./urlSplitter";

export default function bookParser(book:IapiBook):IBook{

  return {
    id:urlSplitter(book.url),
    isbn: book.isbn,
    name: book.name,
    released: book.released,
    numberOfPages: book.numberOfPages,
    //merge all type of characters, just IDs
    characters: [...book.characters, ...book.povCharacters].reduce(
      (previousValue, currentValue) => {
        const 
        characterId = urlSplitter(currentValue),
        isExist = previousValue.find((id) => id === characterId)
        if (!isExist) previousValue.push(characterId)
        return previousValue
      },
      []
    ),
  }
}