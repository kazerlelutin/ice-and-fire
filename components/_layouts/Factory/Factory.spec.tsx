import { afterEach, describe, expect, it, vi } from 'vitest'
import { cleanup } from '@testing-library/react'
import { render, screen } from '../../../utils/test-utils'
import Factory from './Factory'

describe('<Factory />', () => {

  afterEach(cleanup)
  
  it('Check snapshot', async () => {
    
    const useFakeHook = ()=>{
      return  {msg: 'hi !'}
    }

    const FakeComp = ({msg})=>{
      return <p>{msg}</p>
    }
    const FakeCompWithFactory = Factory(useFakeHook,FakeComp);
    const { asFragment } = render(<FakeCompWithFactory/>)

    expect(await screen.findByText('hi !')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })
})
