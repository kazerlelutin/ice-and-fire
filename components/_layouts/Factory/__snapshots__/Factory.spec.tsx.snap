// Vitest Snapshot v1

exports[`<Factory /> > Check snapshot 1`] = `
<DocumentFragment>
  <p>
    hi !
  </p>
</DocumentFragment>
`;
